### PR TITLE
Implement new ExternalContributors plugin

### PR DESCRIPTION
**Pull request by AyodeAwe**
This PR implements a new `ExternalContributors` plugin, which allows RAPIDS organization members to approve PRs from external contributors via a comment before CI is run on them.

### CODE REVIEW

File: src/plugins/ExternalContributors/pr_ex_contibutors.ts [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+    }
+
+
+    private async authorIsNotExternalContributor(author: any, org: any) {
+        return this.context.octokit.orgs.checkMembershipForUser({username: author, org})
+        .then(data => data.status == (204 as any))
+        .catch(_ => false)
+    }
+
+    private async validCommentsExist(): Promise<boolean> {
```
Comment (ajschmidt8) [inline src/plugins/ExternalContributors/pr_ex_contibutors.ts:67]: It seems like this function is very similar to the one below. Can you see if we can consolidate these into a shared function?



- https://github.com/rapidsai/ops-bot/blob/f8e92236a6c0d8de54907ed5e6b465679135090c/src/plugins/AutoMerger/auto_merger.ts#L241-L275

File: src/plugins/ExternalContributors/pr_review_ex_contributors.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
+import { EmitterWebhookEvent } from "@octokit/webhooks";
+import { featureIsDisabled, issueIsPR } from "../../shared";
+import { AutoMergerContext, IssueCommentContext, PRContext, PRReviewContext } from "../../types";
```
Comment (ajschmidt8) [inline src/plugins/ExternalContributors/pr_review_ex_contributors.ts:3]:
```suggestion
import { featureIsDisabled, issueIsPR } from "../../shared";
import { IssueCommentContext } from "../../types";
```

Unused imports

File: src/plugins/ExternalContributors/pr_review_ex_contributors.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+        
+    }
+
+    async pipePR(): Promise<any> {
+        if (await featureIsDisabled(this.context, "external_contributors")) return;
+        const { payload } = this.context
+        const prNumber = payload.issue.number
+        const username = payload.comment.user.login
+
+        if(!["ok to test", "okay to test"].includes(payload.comment.body)) return
```
Comment (ajschmidt8) [inline src/plugins/ExternalContributors/pr_review_ex_contributors.ts:16]: It seems like the `ok to test` and `okay to test` strings exist in both of these files. Can you DRY them up? Maybe move them into a `src/plugins/ExternalContributors/shared.ts` file?

Comment (ajschmidt8) [inline src/plugins/ExternalContributors/pr_review_ex_contributors.ts:16]: It seems like the `ok to test` and `okay to test` strings exist in both of these files. Can you DRY them up? Maybe move them into a `src/plugins/ExternalContributors/shared.ts` file?

File: src/plugins/ExternalContributors/pr_review_ex_contributors.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+        return await this.context.octokit.rest.git.createRef({
+            ref: `refs/heads/external-pr-${payload.issue.number}`,
+            repo: payload.repository.name,
+            owner: payload.repository.owner.login,
+            sha: pr.data.head.sha
+        })
+    }
+
+    private async authorHasPermission(actor) {
+        return ["admin", "write"].includes((
```
Comment (ajschmidt8) [inline src/plugins/ExternalContributors/pr_review_ex_contributors.ts:53]: These permission strings `admin` and `write` also exist in both files.
Can you DRY them up as well? Maybe move them into a `src/plugins/ExternalContributors/shared.ts` file?

File: src/shared.ts [hunk 5]
```diff
@@ -81,3 +81,14 @@ export const featureIsDisabled = async (
...
 };
+
+
+
+/**
+ * Returns true if the payload associated with the provided context
+ * is from a GitHub Pull Request (as opposed to a GitHub Issue).
+ * @param context
+ */
+export const issueIsPR = (context: IssueCommentContext): boolean => {
```
Comment (ajschmidt8) [inline src/shared.ts:92]: This exact same function exists below. Can you DRY them up?

https://github.com/rapidsai/ops-bot/blob/f8e92236a6c0d8de54907ed5e6b465679135090c/src/plugins/AutoMerger/auto_merger.ts#L205-L207

File: src/plugins/ExternalContributors/pr_ex_contibutors.ts [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+            // Update commit on the source repository branch to match forked branch
+            return await this.context.octokit.rest.git.updateRef({
+                ref: `heads/${getExternalPRBranchName(payload.pull_request.number)}`,
+                repo: payload.repository.name,
+                owner: payload.repository.owner.login,
+                sha: payload.pull_request.head.sha,
+                force: true
+            })
+
+        }
```
Comment (ajschmidt8) [inline src/plugins/ExternalContributors/pr_ex_contibutors.ts:45]: In the case where PRs are `reopened`, the `ref`/`branch` would probably already have been deleted, right? So this logic might need to be updated accordingly. Perhaps by making a new `createOrUpdateRef` helper function or something

File: src/plugins/ExternalContributors/index.ts [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+import { PRExternalContributors } from "./pr_ex_contibutors";
+import { PRReviewExternalContributors } from "./pr_review_ex_contributors";
+
+export const initExternalContributors = (app: Probot) => {
+    app.on(
+      [
+        "pull_request.opened",
+        "pull_request.synchronize",
+        "pull_request.closed"
+      ],
```
Comment (ajschmidt8) [inline src/plugins/ExternalContributors/index.ts:11]: probably need to either add a `pull_request.reopened` event to this list or just remove all of these and simplify it to only `pull_request`.

File: src/types.ts [hunk 8]
```diff
@@ -6,7 +6,7 @@ import {
   EmitterWebhookEventName,
 } from "@octokit/webhooks";
 
-type ContextFactory<E extends EmitterWebhookEventName> =
+export type ContextFactory<E extends EmitterWebhookEventName> =
```
Comment (ajschmidt8) [inline src/types.ts:9]:
```suggestion
type ContextFactory<E extends EmitterWebhookEventName> =
```

I don't think this is being used in any other files, right? Can probably be un-exported.

**Review by ajschmidt8 — COMMENTED** [review]
:fire: :fire: :fire: :fire: :fire: looking awesome. few comments left

File: test/external_contributors.test.ts [hunk 9]
```diff
@@ -213,25 +213,27 @@ describe('External Contributors', () => {
             mockPaginate.mockResolvedValueOnce([{body: commentBody, user:{login: "jake"}}])
             mockGetUserPermissionLevel.mockResolvedValueOnce({data: {permission}})
             mockCreateRef.mockResolvedValueOnce(true)
-            mockGetRef.mockResolvedValueOnce({status: 302})
+            mockUpdateRef.mockRejectedValueOnce({status: 302})
```
Comment (ajschmidt8) [inline test/external_contributors.test.ts:216]: It looks like `octokit` is using `node-fetch` under-the-hood` ([src](https://github.com/octokit/request.js/blob/d65eb9ff3412a1490b20506e3d815eddc976d09b/package.json#L29)).

According to the `node-fetch` [docs here](https://github.com/node-fetch/node-fetch/tree/2.x#handling-exceptions), it seems that `3xx` responses are not handled as exceptions. So you might need to adjust this code accordingly.

Comment (AyodeAwe) [reply test/external_contributors.test.ts:216]: Hmm, postman returned a 422 HTTP error when I teste
Screen Shot 2022-03-24 at 12 54 59 PM
d

**Review by ajschmidt8 — APPROVED** [review]
:fire: :fire: :fire:. awesome work! let's get this merged and start testing it out.

**Pull request merged**